transaction-history-manager/src/main/java/com/qpay/transactionhistorymanager/service/impl/TransactionHistoryServiceImpl.java

```diff
@@ -31,4 +47,112 @@ public TransactionEntity saveTransaction(final TransactionModification transacti
         final var transactionEntity = transactionMapper.map(transactionModification);
         return transactionRepository.save(transactionEntity);
     }
-}
+
+    public TransactionOutcome getTransactionsOutcome(final long walletId, final LocalDateTime startDate, final LocalDateTime endDate) {
+        final var income = getIncome(walletId, startDate, endDate);
+        final var expenses = getExpenses(walletId, startDate, endDate);
+        return TransactionOutcome.builder()
+                .income(income)
+                .expenses(expenses)
+                .build();
+    }
+
+    private Outcome getExpenses(final long walletId, final LocalDateTime startDate, final LocalDateTime endDate) {
+        final var filterTransactions = Aggregation.match(
+                Criteria.where("transactionType").is(TransactionType.PAYMENT)
+                        .and("idFrom").is(walletId)
+                        .and("createdAt").gte(startDate).lte(endDate)
+        );
+
+        final var groupByIdTo = Aggregation.group("idTo")
+                .sum("amount").as("amount")
+                .first("nameTo").as("destination");
+
+        final var sortDesc = Aggregation.sort(Sort.by(Sort.Direction.DESC, "amount"));
+
+        final var groupTransactionsAndCalculateTotal = Aggregation.group()
+                .sum("amount").as("total")
+                .push(new BasicDBObject()
+                        .append("idTo", "$_id")
+                        .append("amount", "$amount")
+                        .append("destination", "$destination"))
+                .as("transactionsGroup");
+
+        final var limitTransactionsGroup = Aggregation.project()
+                .andExclude("_id")
+                .andInclude("total")
+                .andExpression("transactionsGroup").slice(3).as("transactionsGroup");
+
+        final var unwindTransactionsGroup = Aggregation.unwind("transactionsGroup");
+
+        final var addShareField = Aggregation.addFields()
+                .addFieldWithValue("transactionsGroup.share",
+                        AggregationExpression.from(
+                                MongoExpression.create("{ $round: { $multiply: [{ $divide: ['$transactionsGroup.amount', '$total'] }, 100] } }")
+                        )
+                ).build();
+
+        final var finalGroupOperation = Aggregation.group()
+                .first("total").as("total")
+                .push("transactionsGroup").as("transactionsGroup");
+
+        final var finalProjection = Aggregation.project()
+                .andExclude("_id");
+
+        final var aggregation = Aggregation.newAggregation(
+                filterTransactions,
+                groupByIdTo,
+                sortDesc,
+                groupTransactionsAndCalculateTotal,
+                limitTransactionsGroup,
+                unwindTransactionsGroup,
+                addShareField,
+                finalGroupOperation,
+                finalProjection
+        );
```

Comment:
I would create a dedicated class, e.g. `AggregationPipelineGenerator` which returns this pipeline `final var aggregation = Aggregation.newAggregation(...`